src/ripple/basics/impl/PerfLogImp.cpp

```diff
@@ -0,0 +1,513 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2018 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/basics/BasicConfig.h>
+#include <ripple/basics/impl/PerfLogImp.h>
+#include <ripple/beast/core/CurrentThreadName.h>
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/json/json_writer.h>
+#include <ripple/json/to_string.h>
+#include <ripple/rpc/impl/Handler.h>
+#include <atomic>
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+namespace ripple {
+namespace perf {
+
+void
+PerfLogImp::Counters::populateRpc(std::vector<char const*> const& labels)
+{
+    for ( auto label : labels)
+    {
+        auto const &elem = rpc_.emplace(label, Rpc(label));
+        if (!elem.second)
+            assert(false);
+    }
+}
+
+void
+PerfLogImp::Counters::populateJq(JobTypes const& jobTypes)
+{
+    for (auto const& job : jobTypes)
+    {
+        auto elem = jq_.emplace(job.first, Jq(job.second.name()));
+        if (!elem.second)
+            assert(false);
+    }
+}
+
+Json::Value
+PerfLogImp::Counters::countersJson() const
+{
+    Json::Value rpcobj(Json::objectValue);
+    // totalRpc represents all rpc methods. All that started, finished, etc.
+    Rpc totalRpc("total");
+    for (auto const& proc : rpc_)
+    {
+        Json::Value p(Json::objectValue);
+        {
+            std::unique_lock<std::mutex> lock(proc.second.mut);
+            auto const tmp = proc.second.sync;
+            lock.unlock();
+            p[jss::started] = std::to_string(tmp.started);
```

Comment:
I glanced through the log this is generating.  So I'll ask a question.  Do you think there might be savings in the file size if you only do this work if `tmp.started != 0`?  It certainly would make a difference at start up.  But it might even make a difference after the server has been running for quite a while.  I have to suspect that many of the RPC entry points never get used in the usual life of a rippled instance.

Just a thought.

Reply:
fixed